MOI.jl

```diff
@@ -0,0 +1,282 @@
+    constraint_dual::Vector{Float64}
+    # Timers
+    initialize_timer::Float64
+    eval_objective_timer::Float64
+    eval_constraint_timer::Float64
+    eval_objective_gradient_timer::Float64
+    eval_constraint_gradient_timer::Float64
+    eval_constraint_jacobian_timer::Float64
+    eval_hessian_objective_timer::Float64
+    eval_hessian_constraint_timer::Float64
+    eval_hessian_lagrangian_timer::Float64
+    parameters_as_variables::Bool
+
+    function MOI.Evaluator(
+        model::Model,
+        backend::B = nothing,
+    ) where {B<:Union{Nothing,MOI.AbstractMathOptInterface.NLPEvaluator}}
+        return new{B}(
+            model,
+            backend,
+            ConstraintIndex[],
+            Float64[],
+            0.0,
+            0.0,
+            0.0,
+            0.0,
+            0.0,
+            0.0,
+            0.0,
+            0.0,
+            0.0,
+            false
+        )
+    end
+end
+
+function Evaluator{B}(
+    # The internal datastructure.
+    model::Model,
+    # The abstract-differentiation backend
+    backend::B,
+    # ordered_constraints is needed because `OrderedDict` doesn't support
+    # looking up a key by the linear index.
+    ordered_constraints::Vector{ConstraintIndex},
+    # Storage for the NLPBlockDual, so that we can query the dual of individual
+    # constraints without needing to query the full vector each time.
+    constraint_dual::Vector{Float64},
+    # Timers
+    initialize_timer::Float64,
+    eval_objective_timer::Float64,
+    eval_constraint_timer::Float64,
+    eval_objective_gradient_timer::Float64,
+    eval_constraint_gradient_timer::Float64,
+    eval_constraint_jacobian_timer::Float64,
+    eval_hessian_objective_timer::Float64,
+    eval_hessian_constraint_timer::Float64,
+    eval_hessian_lagrangian_timer::Float64,
+) where {B<:Union{Nothing,MOI.AbstractMathOptInterface.NLPEvaluator}}
+    return Evaluator{B}(
+        # The internal datastructure.
+        model::Model,
+        # The abstract-differentiation backend
+        backend::B,
+        # ordered_constraints is needed because `OrderedDict` doesn't support
+        # looking up a key by the linear index.
+        ordered_constraints::Vector{ConstraintIndex},
+        # Storage for the NLPBlockDual, so that we can query the dual of individual
+        # constraints without needing to query the full vector each time.
+        constraint_dual::Vector{Float64},
+        # Timers
+        initialize_timer::Float64,
+        eval_objective_timer::Float64,
+        eval_constraint_timer::Float64,
+        eval_objective_gradient_timer::Float64,
+        eval_constraint_gradient_timer::Float64,
+        eval_constraint_jacobian_timer::Float64,
+        eval_hessian_objective_timer::Float64,
+        eval_hessian_constraint_timer::Float64,
+        eval_hessian_lagrangian_timer::Float64,
+        false
+    )
+end
+
+
+function MOI.Nonlinear.ReverseAD._forward_eval_ϵ(
```

Comment:
Just an FYI: we won't be merging this into DiffOpt if you touch private methods in MOI.

I think the "right" way to do this is to reconstruct a new nonlinear model where you explicitly convert the parameters to decision variables.

Reply:
I agree. I was trying to overload a few things to see if I had it right about where the MOI code threats Parameters and Variables different (hess and jacb wise).

Reply:
I will implement the approach where I reconstruct the problem ! :)

Reply:
done, seems that the jacobian is now being constructed considering the parameters :)